site/src/modules/resources/AgentDevcontainerCard.tsx

```diff
@@ -1,191 +1,368 @@
-		let recreateSucceeded = false;
-		try {
+	const { browser_only } = useFeatureVisibility();
+	const { proxy } = useProxy();
+	const queryClient = useQueryClient();
+
+	// The sub agent comes from the workspace response whereas the devcontainer
+	// comes from the agent containers endpoint. We need alignment between the
+	// two, so if the sub agent is not present or the IDs do not match, we
+	// assume it has been removed.
+	const subAgent = subAgents.find((sub) => sub.id === devcontainer.agent?.id);
+
+	const appSections = (subAgent && organizeAgentApps(subAgent.apps)) || [];
+	const displayApps =
+		subAgent?.display_apps.filter((app) => {
+			if (browser_only) {
+				return ["web_terminal", "port_forwarding_helper"].includes(app);
+			}
+			return true;
+		}) || [];
+	const showVSCode =
+		devcontainer.container &&
+		(displayApps.includes("vscode") || displayApps.includes("vscode_insiders"));
+	const hasAppsToDisplay =
+		displayApps.includes("web_terminal") ||
+		showVSCode ||
+		appSections.some((it) => it.apps.length > 0);
+
+	const rebuildDevcontainerMutation = useMutation({
+		mutationFn: async () => {
 			const response = await fetch(
-				`/api/v2/workspaceagents/${agent.id}/containers/devcontainers/container/${container.id}/recreate`,
-				{
-					method: "POST",
-				},
+				`/api/v2/workspaceagents/${parentAgent.id}/containers/devcontainers/container/${devcontainer.container?.id}/recreate`,
+				{ method: "POST" },
 			);
 			if (!response.ok) {
 				const errorData = await response.json().catch(() => ({}));
 				throw new Error(
-					errorData.message || `Failed to recreate: ${response.statusText}`,
+					errorData.message || `Failed to rebuild: ${response.statusText}`,
 				);
 			}
-			// If the request was accepted (e.g. 202), we mark it as succeeded.
-			// Once complete, the component will unmount, so the spinner will
-			// disappear with it.
-			if (response.status === 202) {
-				recreateSucceeded = true;
+			return response;
+		},
+		onMutate: async () => {
+			await queryClient.cancelQueries({
+				queryKey: ["agents", parentAgent.id, "containers"],
+			});
+
+			// Snapshot the previous data for rollback in case of error.
+			const previousData = queryClient.getQueryData([
+				"agents",
+				parentAgent.id,
+				"containers",
+			]);
+
+			// Optimistically update the devcontainer status to
+			// "starting" and zero the agent and container to mimic what
+			// the API does.
+			queryClient.setQueryData(
+				["agents", parentAgent.id, "containers"],
+				(oldData?: WorkspaceAgentListContainersResponse) => {
+					if (!oldData?.devcontainers) return oldData;
+					return {
+						...oldData,
+						devcontainers: oldData.devcontainers.map((dc) => {
+							if (dc.id === devcontainer.id) {
+								return {
+									...dc,
+									agent: null,
+									container: null,
+									status: "starting",
+								};
+							}
+							return dc;
+						}),
+					};
+				},
+			);
+
+			return { previousData };
+		},
```

Comment:
@BrunoQuaresma should I keep or remove this? It essentially just leads to a slightly faster UI response.

Reply:
It is up to you! If you think this faster UI experience, worth the amount of code/complexity, go for it.